nexus/db-model/src/webhook_delivery.rs

```diff
@@ -0,0 +1,201 @@
+            // There's a singleton entry in the `webhook_event` table for
+            // probes, so that we can reference a real event ID but need not
+            // create a bunch of duplicate empty events every time a probe is sent.
+            event_id: WebhookEventUuid::from_untyped_uuid(
+                WebhookEvent::PROBE_EVENT_ID,
+            )
+            .into(),
+            rx_id: (*rx_id).into(),
+            trigger: WebhookDeliveryTrigger::Probe,
+            state: WebhookDeliveryState::Pending,
+            payload: serde_json::json!({}),
+            attempts: SqlU8::new(0),
+            time_created: Utc::now(),
+            time_completed: None,
+            deliverator_id: Some((*deliverator_id).into()),
+            time_delivery_started: Some(Utc::now()),
+        }
+    }
+
+    pub fn to_api_delivery(
+        &self,
+        event_class: WebhookEventClass,
+        attempts: &[WebhookDeliveryAttempt],
+    ) -> views::WebhookDelivery {
+        let mut view = views::WebhookDelivery {
+            id: self.id.into_untyped_uuid(),
+            webhook_id: self.rx_id.into(),
+            event_class: event_class.as_str().to_owned(),
+            event_id: self.event_id.into(),
+            state: self.state.into(),
+            trigger: self.trigger.into(),
+            attempts: attempts
+                .iter()
+                .map(views::WebhookDeliveryAttempt::from)
+                .collect(),
+        };
+        // Make sure attempts are in order; each attempt entry also includes an
+        // attempt number, which should be used authoritatively to determine the
+        // ordering of attempts, but it seems nice to also sort the list,
+        // because we can...
+        view.attempts.sort_by_key(|a| a.attempt);
+        view
+    }
+}
+
+/// An individual delivery attempt for a [`WebhookDelivery`].
+#[derive(
+    Clone,
+    Queryable,
+    Debug,
+    Selectable,
+    Serialize,
+    Deserialize,
+    Insertable,
+    PartialEq,
+)]
+#[diesel(table_name = webhook_delivery_attempt)]
+pub struct WebhookDeliveryAttempt {
+    /// ID of the delivery entry (foreign key into `webhook_delivery`).
+    pub delivery_id: DbTypedUuid<WebhookDeliveryKind>,
+
+    /// Attempt number (retry count).
+    pub attempt: SqlU8,
+
+    /// ID of the receiver to which this event is dispatched (foreign key into
+    /// `webhook_rx`).
+    pub rx_id: DbTypedUuid<WebhookReceiverKind>,
+
+    pub result: WebhookDeliveryAttemptResult,
+
+    pub response_status: Option<i16>,
+
+    #[serde(with = "optional_time_delta")]
+    pub response_duration: Option<TimeDelta>,
+
+    pub time_created: DateTime<Utc>,
+
+    pub deliverator_id: DbTypedUuid<OmicronZoneKind>,
+}
+
+impl WebhookDeliveryAttempt {
+    fn response_view(&self) -> Option<views::WebhookDeliveryResponse> {
+        Some(views::WebhookDeliveryResponse {
+            status: self.response_status? as u16, // i hate that this has to signed in the database...
```

Comment:
You might be interested in https://github.com/oxidecomputer/omicron/blob/main/nexus/db-model/src/unsigned.rs - we have used different types here to make unexpectedly-signed values a serialization error when we try reading them from the database.